### ISSUE TITLE
syntax error, unexpected ']' in excon-0.107.0/lib/excon/socket.rb:213: 

### ISSUE THREAD

**Issue by furan917** [excon/excon#843]
Hey, sorry to bother you, you recently released your 1.107.0 tag and this has caused some of our infrastructure that uses chef to throw the following stacktrace. (I am not a ruby dev sorry) Is this expected behaviour due to a BC and ruby version (I am guessing so) or something that could be fixed?

This caught us off guard but we will be pinning to 1.106 for the areas affected. (This is the danger not pinning to begin with!)

```

Running handlers:
[2023-12-15T15:36:00+00:00] ERROR: Running exception handlers
Running handlers complete
[2023-12-15T15:36:00+00:00] ERROR: Exception handlers complete
Chef Client failed. 0 resources updated in 12 seconds
[2023-12-15T15:36:00+00:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
[2023-12-15T15:36:00+00:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2023-12-15T15:36:00+00:00] FATAL: SyntaxError: /opt/chef/embedded/lib/ruby/gems/2.5.0/gems/excon-0.107.0/lib/excon/socket.rb:84: syntax error, unexpected .., expecting ']'
            result << block[..idx]
                            ^~
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/excon-0.107.0/lib/excon/socket.rb:213: syntax error, unexpected ']'
... = @read_buffer[@read_offset..]
...                              ^
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/excon-0.107.0/lib/excon/socket.rb:292: syntax error, unexpected ']'
        @read_buffer[start..]
                            ^
/opt/chef/embedded/lib/ruby/gems/2.5.0/gems/excon-0.107.0/lib/excon/socket.rb:408: syntax error, unexpected end-of-input, expecting keyword_end
```

**Comment by geemus** [comment]
Sorry about that, I didn't have any reason to believe it was unsafe.

What Ruby version are you using? (the gem version in the stack traces is from 2015)

My suspicion is that this is a syntax change from earlier to later Ruby's, and we may have missed it because we don't run CI against unsupported Ruby versions. I could certainly be wrong though.

**Comment by geemus** [comment]
And as suggested, I would expect pinning to `0.106.0` should be safe and avoid this problem until I can land a fix.

**Comment by jasons42** [comment]
We're seeing a similar issue this morning, not in Chef but in our CI/CD pipeline scripts.

We're running Ruby 2.5.1.

```
/var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/excon-0.107.0/lib/excon.rb:38:in `require': /var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/excon-0.107.0/lib/excon/socket.rb:84: syntax error, unexpected .., expecting ']' (SyntaxError)
            result << block[..idx]
                            ^~
/var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/excon-0.107.0/lib/excon/socket.rb:213: syntax error, unexpected ']'
... = @read_buffer[@read_offset..]
...                              ^
/var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/excon-0.107.0/lib/excon/socket.rb:292: syntax error, unexpected ']'
        @read_buffer[start..]
                            ^
/var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/excon-0.107.0/lib/excon/socket.rb:408: syntax error, unexpected end-of-input, expecting keyword_end
	from /var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/excon-0.107.0/lib/excon.rb:38:in `<top (required)>'
	from /var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/faraday-1.8.0/lib/faraday/dependency_loader.rb:11:in `require'
	from /var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/faraday-1.8.0/lib/faraday/dependency_loader.rb:11:in `dependency'
	from /var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/faraday-excon-1.1.0/lib/faraday/adapter/excon.rb:7:in `<class:Excon>'
	from /var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/faraday-excon-1.1.0/lib/faraday/adapter/excon.rb:6:in `<class:Adapter>'
	from /var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/faraday-excon-1.1.0/lib/faraday/adapter/excon.rb:4:in `<module:Faraday>'
	from /var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/faraday-excon-1.1.0/lib/faraday/adapter/excon.rb:3:in `<top (required)>'
	from /var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/faraday-excon-1.1.0/lib/faraday/excon.rb:3:in `require_relative'
	from /var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/faraday-excon-1.1.0/lib/faraday/excon.rb:3:in `<top (required)>'
	from /var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/faraday-1.8.0/lib/faraday.rb:34:in `require'
	from /var/lib/jenkins/.rbenv/versions/2.5.1/lib/ruby/gems/2.5.0/gems/faraday-1.8.0/lib/faraday.rb:34:in `<top (required)>'
```

**Comment by geemus** [comment]
Ok, that narrows it down. I bet it's a problem with explicit vs implicit arguments, which should be simple to fix (hopefully).

**Comment by davek804** [comment]
Similar stacktrace here, if it helps narrow things down for you:

`Fetching gem metadata from https://rubygems.org/
Fetching version metadata from https://rubygems.org/
Fetching dependency metadata from https://rubygems.org/
Resolving dependencies...
Using chef-vault 3.4.3
Installing excon 0.107.0
Using multi_json 1.15.0
Using bundler 1.12.5
Using docker-api 1.34.2
Bundle complete! 2 Gemfile dependencies, 5 gems now installed.
Use `bundle show [gemname]` to see where a bundled gem is installed.
Compiling Cookbooks...
================================================================================
Recipe Compile Error in /var/chef/cache/cookbooks/docker/libraries/docker_base.rb
================================================================================
SyntaxError
-----------
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/excon-0.107.0/lib/excon/socket.rb:84: syntax error, unexpected .., expecting ']'
            result << block[..idx]
                              ^
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/excon-0.107.0/lib/excon/socket.rb:213: syntax error, unexpected ']'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/excon-0.107.0/lib/excon/socket.rb:292: syntax error, unexpected ']'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/excon-0.107.0/lib/excon/socket.rb:408: syntax error, unexpected end-of-input, expecting keyword_end
Cookbook Trace:
---------------
  /var/chef/cache/cookbooks/docker/libraries/docker_base.rb:3:in `<class:DockerBase>'
  /var/chef/cache/cookbooks/docker/libraries/docker_base.rb:2:in `<module:DockerCookbook>'
  /var/chef/cache/cookbooks/docker/libraries/docker_base.rb:1:in `<top (required)>'
Relevant File Content:
----------------------
/var/chef/cache/cookbooks/docker/libraries/docker_base.rb:
  1:  module DockerCookbook
  2:    class DockerBase < Chef::Resource
  3>>     require 'docker'
  4:      require 'shellwords'
  5:  
  6:      ################
  7:      # Helper methods
  8:      ################
  9:  
 10:      def connection
 11:        @connection ||= begin
 12:                          opts = {}
System Info:
------------
chef_version=12.22.1
platform=centos
platform_version=7.9.2009
ruby=ruby 2.3.6p384 (2017-12-14 revision 61[254](https://gitlab.tools.dtas.co/flag/chef/fd13/-/jobs/227090#L254)) [x86_64-linux]
program_name=chef-client worker: ppid=30851;start=10:59:52;
executable=/opt/chef/bin/chef-client
Running handlers:
[2023-12-15T11:00:38-05:00] ERROR: Running exception handlers
Running handlers complete
[2023-12-15T11:00:38-05:00] ERROR: Exception handlers complete
Chef Client failed. 0 resources updated in 45 seconds
[2023-12-15T11:00:38-05:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
[2023-12-15T11:00:38-05:00] FATAL: Please provide the contents of the stacktrace.out file if you file a bug report
[2023-12-15T11:00:38-05:00] ERROR: /opt/chef/embedded/lib/ruby/gems/2.3.0/gems/excon-0.107.0/lib/excon/socket.rb:84: syntax error, unexpected .., expecting ']'
            result << block[..idx]
                              ^
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/excon-0.107.0/lib/excon/socket.rb:213: syntax error, unexpected ']'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/excon-0.107.0/lib/excon/socket.rb:292: syntax error, unexpected ']'
/opt/chef/embedded/lib/ruby/gems/2.3.0/gems/excon-0.107.0/lib/excon/socket.rb:408: syntax error, unexpected end-of-input, expecting keyword_end
[2023-12-15T11:00:38-05:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)`

**Comment by geemus** [comment]
I think I know the fix, but please bear with me for a bit as I try to verify (I was unable to build those old versions of Ruby on my machine, so going to try and utilize CI for it). Hoping to push a fix soon via #844

**Comment by geemus** [comment]
v0.108.0 is now out including a fix. Apologies for the negative impact.

For those impacted, you may want to update to a newer supported version of Ruby if possible, and barring that it may be safer to pin to a specific excon version. I do my best to have stable/safe minor releases, but also don't test every version of Ruby, so things like this are liable to slip through from time to time (and although this one was easy enough to fix, they might not always be).
